src/preprocessing/passes/daneshvar.cpp

```diff
@@ -0,0 +1,663 @@
+    for (auto& elem : pattern)
+    {
+        if (elem[0] == 'S') // element is a subtree
+        {
+            getPattern(std::stoi(elem.substr(1)), subtreePattern, pat);
+        }
+        else // symbol is a number (color)
+        {
+            pat.push_back(std::stoi(elem));
+        }
+    }
+}
+
+
+
+std::unique_ptr<NodeInfo> getNodeInfo(const Node& n, uint32_t id)
+{
+    std::map<Node, uint32_t> subtreeCache;
+    std::map<std::string, uint32_t> symbolMap;
+    std::map<uint32_t, std::vector<std::string>> subtreePattern;
+    std::map<uint32_t, std::vector<std::string>> symbolList;
+
+    // Fills the above maps and vectors
+    dfs_iterative(n, subtreeCache, symbolMap, subtreePattern, symbolList);
+
+    // Encoding
+    std::string encoding;
+    for (const auto& [id_key, pattern] : subtreePattern) {
+        encoding += std::to_string(id_key) + ":";
+        for (const auto& symbol : pattern) {
+            encoding += symbol + ",";
+        }
+        if (!pattern.empty()) {
+            encoding.pop_back(); // Remove the last comma
+        }
+        encoding += ";";
+    }
+
+    uint32_t treeId = subtreeCache[n];
+
+    // Pattern
+    std::vector<uint32_t> pat;
+    getPattern(treeId, subtreePattern, pat);
+
+    // Symbols
+    std::vector<std::string> symbols = symbolList[treeId];
+
+    // Role
+    std::map<std::string, int32_t> role;
+    for (size_t i = 0; i < symbols.size(); i++) {
+        if (role.find(symbols[i]) == role.end()) {
+            role[symbols[i]] = i;
+        }
+    }
+
+    // Create NodeInfo using std::make_unique and return the unique_ptr
+    auto ni = std::make_unique<NodeInfo>(
+        n, subtreeCache, symbolMap, subtreePattern, encoding, pat, symbols, role, 0, id);
+
+    return ni;
+}
+
+
+
+
+bool sameClass(const NodeInfo& a, const NodeInfo& b) // Check if two nodes have the same encoding and pattern
+{
+    return a.encoding == b.encoding && a.pat == b.pat;
+}
+
+
+int numDigits(int n)
+{
+    if (n == 0)
+    {
+        return 1;
+    }
+    int count = 0;
+    while (n > 0)
+    {
+        n = n / 10;
+        count++;
+    }
+    return count;
+}
+
+
+
+
+Node rename(const Node& n, 
+            std::unordered_map<std::string, Node>& freeVar2node, 
+            std::unordered_map<std::string, Node>& boundVar2node, 
+            NodeManager* nodeManager,
+            PreprocessingPassContext* d_preprocContext)
+{
+    std::unordered_map<Node, Node> normalized;
+    // Stack entries consist of the node and a boolean indicating if it has been visited
+    std::stack<std::pair<Node, bool>> stack;
+    std::stack<std::unordered_map<std::string, Node>> scopeStack;
+    
+    // Initialize a global variable counter for bound variables
+    static int globalVarCounter = 0;
+
+    // Initialize the stack with the root node, not visited
+    stack.push({n, false});
+    // Initialize the scope stack with the global scope
+    scopeStack.push(boundVar2node);
+
+    while (!stack.empty()) {
```

Comment:
This looks like a tree traversal if I don't miss anything. You don't use a cache to determine whether you already visited a node or not, but only a boolean flag that you push onto the stack.

Why not use the normalized map to cache? You want to have do a post-order DFS here, you can `normalized.emplace(current, Node())` when going down, and store the result to `normalized[current]` when going up. Use similar pattern as suggested above.